### ISSUE TITLE
Editor Line Height

### ISSUE THREAD

**Issue by vkbo** [vkbo/novelWriter#654]
> Is it possible to change the line spacing between lines? I haven't found any settings for that yet, maybe I'm overlooking something.
>
> If not, it would be a great addition. When I first started trying to write, I used a line spacing of 1 too, as mostly used by text editors, but I met a professional writer, who asked why I don't write, and even print the manuscript in a line spacing of 1.5! She was surprised I did not!
>
> Apparently, most manuscripts are printed with a line spacing of 1.5 when printed on regular size A4 pages (that's here in Belgium the case) because it reads much easier, which is important for the person evaluating your work, and then the amount of pages corresponds more or less to what it would be when published.
>She recommended me to do the writing with such line spacing too, which I did in other software, and to be honest, it is much better! It is much easier for the eyes.
>
> grtz,
> Bart

_Originally posted by @bartart3d in https://github.com/vkbo/novelWriter/discussions/653_

**Comment by vkbo** [comment]
It's not straight forward in the editor, but need to test:
https://stackoverflow.com/questions/10250533/set-line-spacing-in-qtextedit

**Comment by vkbo** [comment]
The following code will do the trick, but it is far too slow as it has to be applied after the document is loaded sine the `.setPlainText` method resets formatting. For a novel-size reference document (1.2 Mb) the load time goes from ~340 ms to ~30 seconds.

```python
if self.mainConf.lineHeight > 100:
    blockFmt = QTextBlockFormat()
    blockFmt.setLineHeight(self.mainConf.lineHeight, QTextBlockFormat.ProportionalHeight)

    theCursor = self.textCursor()
    theCursor.clearSelection()
    theCursor.select(QTextCursor.Document)
    theCursor.mergeBlockFormat(blockFmt)
```

**Comment by jyhelle** [comment]
I too use this 1.5 line spacing for review copies, but that is part of the magical incantations I perform manually in my cave with my LibreOffice cauldron ... might it be an option for the .odt output ?
Jan

**Comment by vkbo** [comment]
Absolutely. With the ODT converter I wrote last week I have full control of the output as opposed to the current exporter. It will be included in the 1.2 release. I'll make a beta release of it very soon.

**Comment by vkbo** [comment]
Since this will only be available for exports, and that has now been merged to `dev` for 1.2, I'm closing this.

**Comment by HeyMyian** [comment]
Has there maybe been an update since Feb 2021 that would make the load time for a custom line height in the editor more feasible? For certain fonts and font sizes it can feel rather cramped when writing.

**Comment by vkbo** [comment]
No, there are no changes. Qt5, which novelWriter is based on, is no longer being updated with new features. I'm not sure whether this is different in Qt6, but novelWriter is not moving to Qt6 yet.

**Comment by berndschmidl** [comment]
> The following code will do the trick, but it is far too slow as it has to be applied after the document is loaded sine the `.setPlainText` method resets formatting. For a novel-size reference document (1.2 Mb) the load time goes from ~340 ms to ~30 seconds on my fast i7 desktop PC.
> 
> ```python
> if self.mainConf.lineHeight > 100:
>     blockFmt = QTextBlockFormat()
>     blockFmt.setLineHeight(self.mainConf.lineHeight, QTextBlockFormat.ProportionalHeight)
> 
>     theCursor = self.textCursor()
>     theCursor.clearSelection()
>     theCursor.select(QTextCursor.Document)
>     theCursor.mergeBlockFormat(blockFmt)
> ```

In which file would I set this, if I want to try it out?

**Comment by vkbo** [comment]
I can't remember, but somewhere after the `setPlainText` call in the `loadText` function of the `doceditor.py` file.

**Comment by HeyMyian** [comment]
@berndschmidl Did you find where exactly to put the mentioned code, and if you did what were your results?

**Comment by berndschmidl** [comment]
@HeyMyian Sorry, for the late answer. Unfortunately not but I also just tried for 20 minutes or so. Please let me know if you could figure it out.

**Comment by vkbo** [comment]
Yeah, that's just the config value I added for the line height in my test implementation. All you need to do is replace that with the line height value you want.

**Comment by vkbo** [comment]
I just injected this into the code in the function I mentioned earlier (the lines with the green bar) and it worked right away.

![image](https://github.com/vkbo/novelWriter/assets/1619840/59e0cc06-e505-4b89-801c-a2eec4e27168)

**Comment by skofield05** [comment]
@vkbo Thank you! I'm a step further. I think I did exactly what you said in your previous post. But now when I open novelwriter I get this error: 

Environment:
novelWriter Version: 2.0.7
Host OS: linux (6.2.6-76060206-generic)
Python: 3.10.6 (0x30a06f0)
Qt: 5.15.3, PyQt: 5.15.6
lxml: 4.8.0
enchant: 3.2.0

AttributeError:
'Config' object has no attribute 'lineHeight'

Traceback:
  File "/usr/bin/novelwriter", line 33, in <module>
    sys.exit(load_entry_point('novelWriter==2.0.7', 'gui_scripts', 'novelwriter')())

  File "/usr/lib/python3/dist-packages/novelwriter/__init__.py", line 258, in main
    nwGUI.postLaunchTasks(cmdOpen)

  File "/usr/lib/python3/dist-packages/novelwriter/guimain.py", line 354, in postLaunchTasks
    self.showProjectLoadDialog()

  File "/usr/lib/python3/dist-packages/novelwriter/guimain.py", line 893, in showProjectLoadDialog
    self.openProject(dlgProj.openPath)

  File "/usr/lib/python3/dist-packages/novelwriter/guimain.py", line 547, in openProject
    self.openDocument(lastEdited, doScroll=True)

  File "/usr/lib/python3/dist-packages/novelwriter/guimain.py", line 625, in openDocument
    if self.docEditor.loadText(tHandle, tLine):

  File "/usr/lib/python3/dist-packages/novelwriter/gui/doceditor.py", line 396, in loadText
    blockFmt.setLineHeight(self.mainConf.lineHeight, QTextBlockFormat.ProportionalHeight)

**Comment by skofield05** [comment]
Thanks a lot. I missed that. Works fine now!

**Comment by HeyMyian** [comment]
I found a workaround to increase the line height. It uses https://github.com/source-foundry/font-line to change the line height in the font file itself. For simultaneous installations of the same font with different line heights, one will also need https://github.com/chrissimpkins/fontname.py or similar tools to rename the font family.

**Comment by HeyMyian** [comment]
The font-line script itself is a one-line installation with pip. The font family name change is a python file that I somehow had to download and put into the same folder as the font file I was changing -- I have little idea about Python magic but I got this to work very easily.

After that, I didn't bother with any of the reports that font-line can generate and I went straight for the change. It's two lines in the terminal:

`font-line percent 30 Hack-Regular.ttf`
to change the line height

`python3 fontname.py "Hack LG 30" Hack-Regular-linegap30.ttf`
to change the name of the font family.

This has to be repeated for each font file in the family, like regular, italic, and bold.

It works well with the editor itself, but the document header clips the lower parts of the letters, especially with more extreme changes to the line height. I would absolutely prefer a native solution within novelWriter, I kind of see it as a basic setting for any text editor. But if it's not happening, then this seems to be a valid workaround.

Do FAQ already exist for novelWriter? Do you want me to type something up or can you use the above info?

**Comment by vkbo** [comment]
I have made good progress on this issue. Mostly by accident.

The editor in novelWriter is actually a rich text editor forced to run as a plain text editor. There are a few benefits to this, in particular, it's much smoother because the content is laid out with real pixel values, so you can have pixel-by-pixel scrolling.

I recently tested switching to Qt's plain text editor, which allows for no formatting at all, and therefore uses a significantly simpler text layout engine under the hood. The rest of the editor behaves the same. I found this editor setup to be significantly faster without the large text scaling issues of the rich text editor when accessed from Python.

I'm testing this out now in the branch that will eventually be release 2.2 (so not the upcoming release). I have been able to not only successfully apply line height settings other than the default, but also allow full spell checking for very large documents. I've removed the large document limit setting the user can define, and the hard cap of 5 MB of text per document.

The only downside is the lack of smooth scrolling (it scrolls line-by-line), which also affects the "typewriter scrolling" mode. I think it is worth it.

Anyway, I'm therefore re-opening this issue, and scheduling it for the 2.2 Beta release,

**Comment by vkbo** [comment]
Well, that attempt was short lived. The layout engine for plain text ignores the line height setting completely, see [QTBUG-51891](https://bugreports.qt.io/browse/QTBUG-51891). So, until the issue is fixed, we're back to square one here.

I'll leave the issue open though for future reference, and in case it is resolved. If a fix is implemented into a Qt 5.15 LTS release, I'll add the feature.

**Comment by HeyMyian** [comment]
`doceditor.py` seems to have changed a lot. In the recent nW version, where would this snippet go? Or does it also have to be rephrased (for Qt6?)? I don't Python.

```py
blockFmt = QTextBlockFormat()
blockFmt.setLineHeight(self.mainConf.lineHeight, QTextBlockFormat.ProportionalHeight)

theCursor = self.textCursor()
theCursor.clearSelection()
theCursor.select(QTextCursor.Document)
theCursor.mergeBlockFormat(blockFmt)
```

**Comment by vkbo** [comment]
This approach no longer works.

**Comment by vkbo** [comment]
With pull request #2812, it may be possible to go back to using a plain QTextEdit as the base of the editor, which means the QPlainTextEdit is no longer an issue and will no longer block setting line height.

I will queue this for 26.2 release under the assumption this will be possible, but I do need to test it.

**Comment by vkbo** [comment]
@HeyMyian et al.: This is looking promising. I have now reverted the change from release 2.2 that blocked the solution above, so we're back on track. Currently trying to implement this, and then I'll do another alpha release after so it can be tested. I've started pushing alpha releases to pypi as well, so a pipx install can be used for a quick test locally.

With the whole spell checker rewrite, I think I will truncate the milestone for 26.2 soon so we can do a beta test phase without too many new things. These changes touches core features and need proper testing without too much interference from other features.

**Comment by vkbo** [comment]
Heh, while the above fix works for loading the document, there were some real issues keeping the line height setting active while typing. In a rich text editor, like when I worked on the Collett codebase, hitting enter copied the format from the previous line. But novelWriter uses double-enter a lot, which resets to default format.

Thankfully I had a test that caught it, and I put Claude Code on figuring this one out. It took it half an hour to do so, and I pointed it to the Qt source code as well. I'm glad I didn't have to chase this one on my own. But I understand why it failed now. 😅
